junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -70,17 +91,116 @@ public void execute(ExecutionRequest request) {
 	}
 
-	private void executeAllChildren(RunnerExecutor runnerExecutor, TestDescriptor engineDescriptor) {
+	protected CloseableExecutor createExecutor(ExecutionRequest request) {
+		ConfigurationParameters configurationParameters = request.getConfigurationParameters();
+		boolean parallelExecutionEnabled = configurationParameters.getBoolean(PARALLEL_ENABLED_CONFIG).orElse(false);
+		return parallelExecutionEnabled ? new ParallelExecutor(configurationParameters) : new SerialExecutor();
+	}
+
+	private void executeAllChildren(Executor executor, RunnerExecutor runnerExecutor, TestDescriptor engineDescriptor)
+			throws InterruptedException {
+		Collection<? extends TestDescriptor> children = engineDescriptor.getChildren();
+		CountDownLatch latch = new CountDownLatch(children.size());
 		// @formatter:off
-		engineDescriptor.getChildren()
+		children
 				.stream()
 				.map(RunnerTestDescriptor.class::cast)
-				.forEach(runnerExecutor::execute);
+				.map(descriptor -> (Runnable)() -> {
+					runnerExecutor.execute(descriptor);
+					latch.countDown();
+				})
+				.forEach(executor::execute);
 		// @formatter:on
+		latch.await();
+	}
+
+	/**
+	 * Wrapper for {@link Executor} to allow it to be used in a try-with-resources block.
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected interface CloseableExecutor extends Executor, AutoCloseable {
+		default public void close() {
+		};
+	}
+
+	/**
+	 * {@link CloseableExecutor} that executes tasks synchronously.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class SerialExecutor implements CloseableExecutor {
+		public void execute(Runnable command) {
+			command.run();
+		}
+
+	}
+
+	/**
+	 * {@link CloseableExecutor} backed by a {@link ForkJoinPool} that executes tasks asynchronously
+	 * based on the settings in {@value #PARALLEL_CONFIG}. Clients *must* implement their own logic
+	 * to wait for tasks to complete.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class ParallelExecutor implements CloseableExecutor {
+
+		private final ExecutorService executorService;
+
+		/**
+		 * @param parameters test execution configuration
+		 */
+		public ParallelExecutor(ConfigurationParameters parameters) {
+			ConfigurationParameters prefixedParameters = new PrefixedConfigurationParameters(parameters,
+				PARALLEL_CONFIG);
+			String strategyName = prefixedParameters.get("strategy").orElse("dynamic");
+			ParallelExecutionConfigurationStrategy executionStrategy = DefaultParallelExecutionConfigurationStrategy.valueOf(
+				strategyName.toUpperCase());
+			ParallelExecutionConfiguration executionConfiguration = executionStrategy.createConfiguration(
+				prefixedParameters);
+			executorService = createExecutorService(executionConfiguration);
+		}
+
+		public void execute(Runnable command) {
+			executorService.execute(command);
+		}
+
+		public void close() {
+			executorService.shutdown();
```

Comment:
I think we should still await termination with a timeout to know if this behaves differently than we expect. Let's log a warning if it isn't terminated if the timeout expires.

Reply:
I added back in the `awaitTermination` and added warnings if it times out or is interrupted.